### ISSUE TITLE
[lit/reactive-element] docs for queryAssignedNodes are wrong

### ISSUE THREAD

**Issue by ChristianUlbrich** [lit/lit#2171]
# Description
`queryAssignedNodes` is a _"clever"_ wrapper around [assignedNodes()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedNodes#return_value), that always returns a `Node[]`; this is fundamentally different than a `NodeListOf<HTMLElement>` that is currently documented in the JSDocs:
  * it is a real™  `Array`, there is no need for `Array.from()`, because of the mis-leading documentation, we have a bunch of this conversions in the code
  * if the somewhat undocumented feature of querying for the _default slot_ by passing a _falsy_ value as _slotName_, you can get _TextNodes_. They are also fundamentally different than `HTMLElement` and will easily break your code:
```
// ...

@queryAssignedNodes(undefined)
private _defaultSlotNodes: NodeListOf<HTMLElement>;

private _handleSlotChange() {
  this._defaultSlotNodes.forEach(elm => elm.innerHTML = 'Oops');
}

**Comment by AndrewJakubowicz** [comment]
Thanks for the issue! I agree that the documentation is misleading as there are two behaviors, however I don't think the fix is to document based on the `slotName` passed in. Will expand on this statement to brainstorm a resolution.

`slotName` is used in the decorator logic to select the correct HTMLSlotElement. `slotName` will not impact the return type of this decorator.

We then call `assignedNodes()` on the slot which returns a real array (as you mentioned) of `Node`s. In both named and unnamed slots at this point the return type is `Node[]`.

The confusion that isn't explained in the docs is in the `selector` argument. The third arg. It causes this logic to run:

```ts
        if (selector) {
          nodes = nodes.filter(
            (node) =>
              node.nodeType === Node.ELEMENT_NODE &&
              (node as Element).matches(selector)
          );
        }
```

Thus if the third `selector` argument is included. Even a CSS selector filter which selects for everything, i.e. `"*"`, will narrow the return type to `Element[]` instead of `Node[]`.

I believe the cases are:

 - `@queryAssignedNodes(undefined): Node[]`
 - `@queryAssignedNodes(undefined, false, '*'): Element[]`


With `@queryAssignedElements` being introduced which more naturally selects for `Element[]` in all cases, it may be worth considering deprecating this third argument. In the meantime a solution may be to call out that in all cases a real array is returned, and if a selector is included then the return type is automatically narrowed to only include Elements.

cc/ @justinfagnani

**Comment by ChristianUlbrich** [comment]
I think the idea of the documentation was, that whenever you are using _non-default_ slots you will get `HTMLElement[]`, because you cannot _slot_ TextNodes or any Non-ElementNode in _non-default_ slots. Crazily enough you _can_ get TextNodes, but only if it is _fallback_ content **and** if you pass `{flatten: true}` to `assignedNodes()` (WTF?)

It has to be decided, whether the _suggested_ type signature is one, that is always correct™ or if the suggestion should revolve around  a more specific type signature, that is more practical™. The current docs, tried the later but failed.

After having given this some more thought I basically agree with you observation and thus the documentation should be something like:
```markdown
Note, the type of this property should be annotated as `Node[]` if you do not pass a _selector_ and `HTMLElement[]` if you pass one.
```
As I've stated above my _main concern_ is the usage of `NodeListOf<T>`, our junior devs sprinkled the code with an abundance of `Array.from()` calls (which of course happily works with an array too).

Would this be a better idea @justinfagnani  ?

On a side note, I found the documentation on [assignedNodes()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedNodes) _mis-leading_ as well; to me it was not clear that you **have** to pass `{flatten: true}` to actually get _fallback content_. Seems like I need to improve the docs on MDN also. :)

**Comment by justinfagnani** [comment]
`NodeListOf<HTMLElement>` is clearly wrong and should be changed to `Array<Node>` (or possibly `Array<ChildNode>`). And as in @AndrewJakubowicz's second case, it should be `Array<HTMLElement>` if there's any selector.

Regarding Text nodes: I believe you can also get text nodes into named slots if `flatten` is true and an unnamed slot is forwarded to that named slot.

**Comment by AndrewJakubowicz** [comment]
This should be resolved in Lit 2.1. We've introduced a new decorator `queryAssignedElements` which has a consistent return type of `HTMLElement[]` when used with a `selector`. The new `queryAssignedNodes` API is scoped to only returning `Node[]`.
